File: src/main/java/com/tms/csp/util/DynComplex.java
package com.tms.csp.util;

import com.google.common.collect.ImmutableList;
import com.google.common.collect.Iterators;
import com.tms.csp.ExpItTo;
import com.tms.csp.argBuilder.IArgBuilder;
import com.tms.csp.ast.*;
import com.tms.csp.fm.dnnf.products.Cube;
import com.tms.csp.util.ints.IndexedEntry;
import com.tms.csp.util.ints.TreeSequence;
import com.tms.csp.util.it.ExpFilterIterator;
import com.tms.csp.util.varSets.VarSet;
import kotlin.sequences.Sequence;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;


public class DynComplex implements IArgBuilder, PLConstants, Iterable<Exp> {

    private final Space sp;

    //For And
    //    null:   constantFalse
    //    empty:  constantTrue
    public TreeSequence<Exp> args = new TreeSequence<Exp>();
    private VarSet vars;

    public Boolean fcc;


    public DynComplex(Space sp, VarSet vars) {
        this.sp = sp;
        this.vars = vars;
    }

    public DynComplex(Space sp) {
        this(sp, (VarSet) null);
    }


    public DynComplex(Space sp, Iterable<Exp> args) {
        this(sp);
        addAll(args);
    }

    public DynComplex(Space sp, Exp[] args) {
        this(sp);
        addAll(args);
    }

    public DynComplex(DynComplex that) {
        this(that.sp);
        this.args = that.args.copy();
        vars = that.vars.copy();
        fcc = that.fcc;
    }

    public DynComplex copy() {
        return new DynComplex(this);
    }


    public void addAll(Exp[] args) {
        for (Exp arg : args) {
            add(arg);
        }
    }

    public void addAll(Iterable<Exp> args) {
        for (Exp arg : args) {
            add(arg);
        }
    }

    public boolean assertVars() {
        return assertVars("");
    }

    public String assertVarsMsg(VarSet vars1, VarSet vars2, String constraintJustAdded) {
        return "\n  vars[" + vars1 + "]\n  computed[" + vars2 + "]\n  constraint[" + constraintJustAdded + "]";
    }

    private boolean assertVars(String constraintJustAdded) {
        VarSet vars1 = vars;
        VarSet vars2 = _computeVars();

        if (vars1 == null) {
            throw new IllegalStateException();
        }

        boolean chk = vars1.equals(vars2);

        assert chk : assertVarsMsg(vars1, vars2, constraintJustAdded);
        return true;
    }

    public boolean add(Exp e) {
        if (e.isAnd()) throw new IllegalArgumentException("Ands should be added at the csp level");
        Exp old = args.put(e.getExpId(), e);
        boolean ch = (old == null);
        boolean retVal;
        if (!ch) {
            assert old.expId == e.expId;
//            System.err.println("Put Error:");
//            System.err.println("  new: " + e.getExpId() + ": " + e);
//            System.err.println("  old: " + old.getExpId() + ": " + old);
//            boolean sameSpace1 = e.getSpace() == old.getSpace();
//            boolean sameSpace2 = getSpace() == old.getSpace();
//            System.err.println("  sameSpace1[" + sameSpace1 + "]");
//            System.err.println("  sameSpace2[" + sameSpace2 + "]");
//            getSpace().check();

//            System.err.println("Dup complex constraint: " + old);
            retVal = false;
        } else {
            addVars(e);
            retVal = true;
        }

//        if (retVal) {
//            assertVars(e.toString());
//        }
        return retVal;


    }

    private void addVars(Exp e) {
        if (vars == null) {
            vars = getSpace().varSetBuilder();
        }
        this.vars.addVars(e.getVars());
    }


    public int size() {
        return args.size();
    }

    public boolean isEmpty() {
        return args.isEmpty();
    }


    /**
     * @return an iterator over this.args, formula the increasing order of ids.
     */
    public Iterator<Exp> iterator() {
        return argIterator();
    }

    public Iterator<Exp> argIterator() {
        final Iterator<IndexedEntry<Exp>> iter = args.iterator();
        return new Iterator<Exp>() {
            public boolean hasNext() {
                return iter.hasNext();
            }

            public Exp next() {
                return iter.next().value();
            }

            public void remove() {
                throw new UnsupportedOperationException();
            }
        };
    }

    public int computeHash() {
        int sum = 0;
        for (Exp arg : this) {
            sum += arg.hashCode();
        }
        return sum;
    }

    public boolean remove(Exp arg) {
        int expId = arg.getExpId();
        Exp removed = args.remove(expId);
        boolean ch = removed != null;
        if (ch) {
            vars = sp.mkEmptyVarSet();
        }
        return ch;
    }

    public boolean removeAll(Iterable<Exp> complex) {
        boolean ch = false;
        for (Exp exp : complex) {
            if (remove(exp)) {
                ch = true;
            }
        }
        return ch;
    }


    public Exp[] toArray() {
        Exp[] a = new Exp[size()];
        int index = 0;
        for (Exp arg : this) {
            a[index] = arg;
            index++;
        }
        return a;
    }

    public List<Exp> toList() {
        ArrayList<Exp> a = new ArrayList<Exp>(size());
        for (Exp arg : this) {
            a.add(arg);
        }
        return a;
    }

    public int[] toIntArray() {
        int[] a = new int[size()];
        int i = 0;
        for (Exp exp : this) {
            a[i] = exp.getExpId();
        }
        return a;
    }

    public Exp first() {
        return iterator().next();
    }

    public Space getSpace() {
        return first().getSpace();
    }

    public boolean eq(DynComplex that) {
        return args.equals(that.args);
    }

    public boolean contains(Exp e) {
        return args.containsIndex(e.expId);
    }

    public Exp get(int expId) {
        if (args == null) {
            throw new UnsupportedOperationException();
        }
        return args.get(expId);
    }

    private VarSet _computeVars() {
        if (isEmpty()) {
            return sp.mkEmptyVarSet();
        } else {
            VarSet vrs = getSpace().newMutableVarSet();
            for (Exp constraint : this) {
                vrs.addVars(constraint.getVars());
            }
            return vrs;
        }
    }

    public VarSet vars() {
        assert vars != null;
        return vars;
    }


    public boolean remove(Object o) {
        return remove((Exp) o);
    }

    public void clear() {
        vars = sp.mkEmptyVarSet();
        args.clear();
    }

    public boolean isConstantTrue() {
        return args.isEmpty();
    }

    public boolean isConstantFalse() {
        return args == null;
    }

    public int getConstraintCount() {
        return size();
    }


    public Exp toDnnf() {
        if (isEmpty()) return sp.mkTrue();
        Exp exp = mkExp();
        Exp dnnf = exp.toDnnf();

        assert dnnf.isDnnf();
//        assert dnnf.checkDnnf();
        return dnnf;
    }

    @NotNull
    @Override
    public Exp mk() {
        return mkExp();
    }

    public Exp mkExp() {
        if (isConstantFalse()) {
            return sp.mkConstantFalse();
        } else if (isConstantTrue()) {
            return sp.mkConstantTrue();
        } else if (size() == 1) {
            Exp arg = getFirstArg();
            assert arg.getSpace() == sp;
            return arg;
        } else {
            assert size() > 1;
            return sp.mkPosComplex(this);
//            return new ExpSet(space, this).mkExp();
        }
    }

    public Exp getFirstArg() {
        return iterator().next();
    }


    @Override
    public Boolean isFcc() {
        return fcc;
    }

    public Iterable<Exp> vvIt() {
        return new Iterable<Exp>() {
            @Override
            public Iterator<Exp> iterator() {
                return vvIterator();
            }
        };
    }

    public Iterator<Exp> vvIterator() {
        if (isEmpty()) return Iterators.emptyIterator();

        return new ExpFilterIterator(argIterator()) {
            @Override
            public boolean accept(Exp e) {
                return e.isVv();
            }
        };
    }

    public Iterable<Exp> vvpIt() {
        return new Iterable<Exp>() {
            @Override
            public Iterator<Exp> iterator() {
                return vvpIterator();
            }
        };
    }

    public Iterator<Exp> vvpIterator() {
        return new ExpFilterIterator(argIterator()) {
            @Override
            public boolean accept(Exp e) {
                return e.isVVPlus();
            }
        };
    }

    public Exp mkFormula() {
        return mkExp();
    }

    @Override
    public String toString() {
        ImmutableList<Exp> args = ImmutableList.copyOf(argIterator());
        Ser a = new Ser();
        Exp.serializeArgList(a, args);
        return a.toString();
    }


    @Override
    public int getSize() {
        return size();
    }

    @NotNull
    @Override
    public Op getOp() {
        return Op.Formula;
    }

    @NotNull
    @Override
    public Op1 getOp1() {
        return getOp().getOp1();
    }

    @NotNull
    @Override
    public Iterable<Exp> getArgIt() {
        return this::argIterator;
    }


    public boolean containsVar(Var var) {
        return vars.containsVar(var);
    }

    public boolean anyVarOverlap(Lit lit) {
        return vars.anyVarOverlap(lit);
    }

    public boolean anyVarOverlap(Cube cube) {
        return vars.anyVarOverlap(cube);
    }

    public boolean isVarDisjoint(DynCube simple) {
        return vars().isVarDisjoint(simple);
    }

    @NotNull
    public VarSet intersection(@NotNull DynCube simple) {
        return vars().intersection(simple.getVars());
    }

    public Sequence<Exp> asSeq() {
        return ExpItTo.expSeq(getArgIt());
    }

    @NotNull
    @Override
    public Exp[] createExpArray() {
        return ExpFactory.createExpArray(size(), getArgIt());
    }

}

File: src/main/java/com/tms/csp/util/DynComplex.kt
package com.tms.csp.util

import com.google.common.collect.Iterators
import com.tms.csp.argBuilder.IArgBuilder
import com.tms.csp.ast.*
import com.tms.csp.fm.dnnf.products.Cube
import com.tms.csp.util.ints.TreeSequence
import com.tms.csp.util.it.ExpFilterIterator
import com.tms.csp.util.varSets.VarSet


class DynComplex constructor(val space: Space) : IArgBuilder, PLConstants, Iterable<Exp> {

    //For And
    //    null:   constantFalse
    //    empty:  constantTrue
    var args: TreeSequence<Exp>? = null
    var _vars: VarSet? = null

    override var isFcc: Boolean? = null

    constructor(sp: Space, args: Iterable<Exp>) : this(sp) {
        addExpIt(args)
    }

    constructor(sp: Space, args: Array<Exp>) : this(sp) {
        addExpArray(args)
    }

    constructor(that: DynComplex) : this(that.space) {
        this.args = that.args?.copy()
        _vars = that._vars?.copy()
        isFcc = that.isFcc
    }

    val isEmpty: Boolean
        get() = args == null || args!!.isEmpty

    val isConstantTrue: Boolean
        get() = isEmpty


    val firstArg: Exp
        get() = args!!.first().value()

    fun first(): Exp = firstArg

    override val size: Int get() = if (args == null) 0 else args!!.size()

    override val op: Op
        get() = Op.Formula

    override val op1: Op1
        get() = op.op1

    override val argIt: Iterable<Exp>
        get() = Iterable { this.argIter() }

    override fun iterator(): Iterator<Exp> = argIter()


    fun copy(): DynComplex {
        return DynComplex(this)
    }


    fun addExpArray(args: Array<Exp>) {
        for (arg in args) {
            add(arg)
        }
    }

    fun addExpIt(args: Iterable<Exp>) {
        for (arg in args) {
            add(arg)
        }
    }

    fun assertVars(): Boolean {
        return assertVars("")
    }

    fun assertVarsMsg(vars1: VarSet, vars2: VarSet, constraintJustAdded: String): String {
        return "\n  _vars[$vars1]\n  computed[$vars2]\n  constraint[$constraintJustAdded]"
    }

    private fun assertVars(constraintJustAdded: String): Boolean {
        val vars1 = _vars
        val vars2 = _computeVars()

        if (vars1 == null) {
            throw IllegalStateException()
        }

        val chk = vars1 == vars2

        assert(chk) { assertVarsMsg(vars1, vars2, constraintJustAdded) }
        return true
    }

    private val mkArgs: TreeSequence<Exp>
        get() {
            args = args ?: TreeSequence()
            return args!!
        }

    fun add(e: Exp): Boolean {
        requireNotNull(e)
        if (e.isAnd) throw IllegalArgumentException("Ands should be added at the csp level")

        val old = mkArgs.put(e.getExpId(), e)
        val ch = old == null
        return if (!ch) {
            assert(old!!.expId == e.expId)
            //            System.err.println("Put Error:");
            //            System.err.println("  new: " + e.getExpId() + ": " + e);
            //            System.err.println("  old: " + old.getExpId() + ": " + old);
            //            boolean sameSpace1 = e.getSpace() == old.getSpace();
            //            boolean sameSpace2 = getSpace() == old.getSpace();
            //            System.err.println("  sameSpace1[" + sameSpace1 + "]");
            //            System.err.println("  sameSpace2[" + sameSpace2 + "]");
            //            getSpace().check();

            //            System.err.println("Dup complex constraint: " + old);
            false
        } else {
            addVars(e)
            true
        }


    }

    private fun addVars(e: Exp) {
        this._vars = VarSet.union(space, _vars, e.vars);
    }


//    /**
//     * @return an iterator over this.args, formula the increasing order of ids.
//     */
//    override fun iterator(): Iterator<Exp> = expIter()

    fun expIter(): Iterator<Exp> = args.expIter()

    fun argIter(): Iterator<Exp> = args.expIter()

    fun computeHash(): Int {
        var sum = 0
        for (arg in argIt) {
            sum += arg.hashCode()
        }
        return sum
    }

    private val a: TreeSequence<Exp> get() = args!!

    fun remove(arg: Exp): Boolean {
        return if (args == null) {
            false
        } else {
            val expId = arg.getExpId()
            val removed: Exp? = args!!.remove(expId)
            val ch = removed != null
            if (ch) {
                _vars = null
            }
            return ch
        }

    }

    fun removeAll(complex: Iterable<Exp>): Boolean {
        var ch = false
        for (exp in complex) {
            if (remove(exp)) {
                ch = true
            }
        }
        return ch
    }


    fun eq(that: DynComplex): Boolean {
        return args == that.args
    }

    operator fun contains(e: Exp): Boolean {
        return args!!.containsIndex(e.expId)
    }

    operator fun get(expId: Int): Exp? {
        if (args == null) {
            throw UnsupportedOperationException()
        }
        return args!!.get(expId)
    }

    private fun _computeVars(): VarSet {
        return if (isEmpty) {
            space.mkEmptyVarSet()
        } else {
            space.varSetBuilder().apply { argIt.forEach { addVarSet(it.vars) } }.build()
        }
    }

    val vars: VarSet
        get() {
            if (_vars == null) {
                _vars = _computeVars();
            }
            return _vars!!
        }


    fun remove(o: Any): Boolean {
        return remove(o as Exp)
    }

    fun clear() {
        _vars = null
        args = null
    }


    fun toDnnf(): Exp {
        if (isEmpty) return space.mkTrue()
        val exp = mkExp()
        val dnnf = exp.toDnnf()

        assert(dnnf.isDnnf)
        //        assert dnnf.checkDnnf();
        return dnnf
    }


    fun vvIt(): Iterable<Exp> {
        return object : Iterable<Exp> {
            override fun iterator(): Iterator<Exp> {
                return vvIterator()
            }
        }
    }

    fun vvIterator(): Iterator<Exp> {
        return if (isEmpty) Iterators.emptyIterator() else object : ExpFilterIterator(argIter()) {
            override fun accept(e: Exp): Boolean {
                return e.isVv
            }
        }

    }

    fun vvpIt(): Iterable<Exp> {
        return object : Iterable<Exp> {
            override fun iterator(): Iterator<Exp> {
                return vvpIterator()
            }
        }
    }

    fun vvpIterator(): Iterator<Exp> {
        return object : ExpFilterIterator(argIter()) {
            override fun accept(e: Exp): Boolean {
                return e.isVVPlus
            }
        }
    }

    override fun mk(): Exp = when {
        isEmpty -> space.mkConstantTrue()
        size == 1 -> firstArg
        else -> space.mkPosComplex(this)
    }

    fun mkExp(): Exp = mk()

    fun mkFormula(): Exp = mk()

    override fun toString(): String {
        val a = Ser()
        Exp.serializeArgList(a, argIt)
        return a.toString()
    }


    fun containsVar(vr: Var): Boolean {
        return _vars?.containsVar(vr) ?: false
    }

    fun anyVarOverlap(lit: Lit): Boolean {
        return _vars!!.anyVarOverlap(lit)
    }

    fun anyVarOverlap(cube: Cube): Boolean {
        return _vars!!.anyVarOverlap(cube)
    }

    fun isVarDisjoint(simple: DynCube): Boolean {
        return vars.isVarDisjoint(simple)
    }

    fun varOverlap(simple: DynCube): VarSet = vars.overlap(simple.vars)

    fun _asSeq(): Sequence<Exp> = argIt.asSequence()

    override fun createExpArray(): Array<Exp> {
        check(!isEmpty)
        check(size != 1)
        val a = args!!
        val aa: Array<Exp?> = arrayOfNulls(size)
        for ((i, arg) in argIt.withIndex()) {
            aa[i] = arg
        }
        return aa.requireNoNulls()
    }

    fun toArray(): Array<Exp> = createExpArray()

    fun toIntArray(): IntArray {
        check(!isEmpty)
        check(size != 1)
        val aa: IntArray = IntArray(size)
        for (i in 0 until size) aa[i] = args!![i].expId
        return aa
    }


    fun toList(): List<Exp> = argIt.toList()


}

fun DynComplex?.asSeq(): Sequence<Exp> {
    return this?._asSeq() ?: emptySequence()
}